spec.md

```diff
@@ -1090,6 +1084,252 @@ message ControllerServiceCapability {
+// Note that CreateSnapshot no longer blocks after the snapshot is cut.
+// Alternatively, another RPC ListSnapshots can be called with
+// snapshot_id as filtering repeatedly to wait for the upload to
+// complete. ListSnapshots should return immediately with the current
+// snapshot info. When upload is complete, the status of the snapshot
+// from ListSnapshots should become `READY`.
+message SnapshotStatus {
+  enum Type {
+     UNKNOWN = 0;
+     // A snapshot is cut and is now being uploaded.
+     UPLOADING = 1;
+     // A snapshot is ready for use.
+     READY = 2;
+     // A snapshot is in error status.
+     ERROR = 3;
+  }
+  Type type = 1;
+}
+```
+
+##### CreateSnapshot Errors
+
+If the plugin is unable to complete the CreateSnapshot call successfully, it MUST return a non-ok gRPC code in the gRPC status.
+If the conditions defined below are encountered, the plugin MUST return the specified gRPC error code.
+The CO MUST implement the specified error recovery behavior when it encounters the gRPC error code.
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-----------|-----------|-------------|-------------------|
+| Snapshot already exists but is incompatible | 6 ALREADY_EXISTS | Indicates that a snapshot corresponding to the specified snapshot `name` already exists but is incompatible with the specified `volume_id`. | Caller MUST fix the arguments or use a different `name` before retrying. |
+| Operation pending for snapshot | 10 ABORTED | Indicates that there is a already an operation pending for the specified snapshot. In general the Cluster Orchestrator (CO) is responsible for ensuring that there is no more than one call "in-flight" per snapshot at a given time. However, in some circumstances, the CO MAY lose state (for example when the CO crashes and restarts), and MAY issue multiple calls simultaneously for the same snapshot. The Plugin, SHOULD handle this as gracefully as possible, and MAY return this error code to reject secondary calls. | Caller SHOULD ensure that there are no other calls pending for the specified snapshot, and then retry with exponential back off. |
+| Call not implemented | 12 UNIMPLEMENTED | CreateSnapshot call is not implemented by the plugin or disabled in the Plugin's current mode of operation. | Caller MUST NOT retry. Caller MAY call `ControllerGetCapabilities` to discover Plugin capabilities. |
+| Not enough space to create snapshot | 13 RESOURCE_EXHAUSTED | There is not enough space on the storage system to handle the create snapshot request. | Caller should fail this request. Future calls to CreateSnapshot may succeed if space is freed up. |
+
+
+#### `DeleteSnapshot`
+
+A Controller Plugin MUST implement this RPC call if it has CREATE_DELETE_SNAPSHOT capability.
+This RPC will be called by the CO to delete a snapshot.
+If successful, the storage space associated with the snapshot MUST be released and all the data in the snapshot SHALL NOT be accessible anymore.
```

Comment:
"... storage space associated with the snapshot MUST be released..." - this probably depends on how the storage backend reclaims storage space. is there a reason that we feel the need to declare this in the spec? for example, some backend systems might tombstone a snapshot, but it might not actually be deleted until some reaper (or compactor, or whatever) comes along and removes the data.

I see that we have similar language for `DeleteVolume`, perhaps we should create a ticket and address this language in a follow-up PR instead of in this one.

Reply:
Yes, I added this because it is also in DeleteVolume.  I'll create a ticket.

Reply:
https://github.com/container-storage-interface/spec/issues/225